rollouts/controllers/rollout_controller_test.go

```diff
@@ -371,3 +377,158 @@ var _ = Describe("Rollout", func() {
 		})
 	})
 })
+
+func TestReconcileRollout(t *testing.T) {
+	t.Run("maxconcurrent", func(t *testing.T) {
+		// This tests that if "maxConcurrent" is set to 1, only one cluster is synced at
+		// a time. This test calls `reconcileRollout` twice. After the first call,
+		// we check that only the first cluster starts progressing while the second is
+		// waiting. Then, we manually update the sync status of the first remotesync
+		// to "Synced", call `reconcileRollout` again, and verify that the second cluster
+		// starts progressing.
+
+		rollout := &gitopsv1alpha1.Rollout{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Rollout",
+				APIVersion: "gitops.kpt.dev/v1alpha1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "sample",
+			},
+
+			Spec: gitopsv1alpha1.RolloutSpec{
+				PackageToTargetMatcher: gitopsv1alpha1.PackageToClusterMatcher{
+					Type: "AllClusters",
+				},
+				Strategy: gitopsv1alpha1.RolloutStrategy{
+					Type: "RollingUpdate",
+					RollingUpdate: &gitopsv1alpha1.StrategyRollingUpdate{
+						MaxConcurrent: 1,
+					},
+				},
+				Targets: gitopsv1alpha1.ClusterTargetSelector{
+					Selector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"foo": "bar"},
+					},
+				},
+			},
+		}
+
+		// create 2 target clusters with arbitrary names
+		targetClusters := make([]clusterstore.Cluster, 2)
+		for i := 0; i < 2; i++ {
+			targetClusters[i] = clusterstore.Cluster{
+				Ref:    gitopsv1alpha1.ClusterRef{Name: fmt.Sprintf("foo/%d", i)},
+				Labels: map[string]string{"foo": "bar"},
+			}
+		}
+
+		// create an arbitrary package
+		testURL := "https://test.com/git"
+		discoveredPackage := packagediscovery.DiscoveredPackage{
+			Directory: "dir",
+			Revision:  "v0",
+			Branch:    "main",
+			GitHubRepo: &github.Repository{
+				CloneURL: &testURL,
+			},
+		}
+
+		fc := &fakeRolloutsClient{}
+		reconciler := (&RolloutReconciler{Client: fc})
+
+		strategy, err := reconciler.getStrategy(context.Background(), rollout)
+		require.NoError(t, err)
+
+		// first call to reconcileRollout - only one cluster should start progressing
+		_, waveStatus, err := reconciler.reconcileRollout(
+			context.Background(),
+			rollout,
+			strategy,
+			targetClusters,
+			[]packagediscovery.DiscoveredPackage{discoveredPackage},
+		)
+		require.NoError(t, err)
```

Comment:
Are we using this style of assertions in other parts of `kpt` project ? otherwise I have found [omega matchers](https://onsi.github.io/gomega/) to be very good for this.

Reply:
There are a number of places in kpt repo that we are using the `testify/assert` and `testify/require` packages to do similar assertions in unit tests. `require` is a subtly different flavor of `assert`, except that `require` exits the test immediately, while `assert` continues running the rest of the test even when it encounters an error. 

That said, I am happy to convert to omega matchers if you prefer!

Reply:
Didn't see usage of `require` before. But the failure behavior of `require` does make more sense.

I am fine using `require` here, so let's ship it.